javascript-review.js

```diff
@@ -1,92 +1,190 @@
 
-// 3.
+// function Animal(name){
+//     // Object.create(Animal); //you only use Object.create() if you want to inherit the methods and attributes from another object
+//     let animal = {};
+//     animal.name = name;
+//     animal.color = 'red';
+//     return animal;
+// }
+
+// let dog = new Animal('dog');
+// console.log(dog);
+
+// let cat = new Animal('cat');
+// console.log(cat);
+
 
+
+// 3. 
 // Can you access variables defined outside of functions within the scope of functions in JS? Show a simple example of this being used? (Define a constant in global scope and use that in a very simple function.)
 
-// 4.
+//Yes you can, you just need to define the variable you want to use outside the functions as a global variable
+// const y = 10;
+
+// const printXAndY = () => {
+//     let x = 20;
+//     console.log(x, y);
+// }
+
+// printXAndY();
 
+
+// 4.
 // What two ways can be used to access the properties of objects in JS? Give examples of both (although assume the objects have been defined).
+// let obj = {
+//     name: 'Jov',
+//     age: '22',
+//     address: 'Yarrambat'
+// }
+
+// console.log(keys(obj));
+
+// you can iterate through and get the objects keys in an array or access them directly if you know them
 
-// 5.
 
 // What is JSON? Be clear but concise.
+// 5. JSON stands for JavaScript Object Notation and is a format for storing data in and object like fashion, where the keys are all strings.
 
-// 6.
 
 // What is a callback?
+// 6. A callback is a call to another function within a function. Callback functions get passed into other functions that utilise them.
+
 
-// 7.
 
 // Define a function with that takes two number arguments and a callback. The function will add the two numbers and pass them to the callback as an argument. Now call this function and in the callback simply console.log out the argument that has been passed through.
 
-// 8.
+// 8. 
+// const logNumber = (n) => {console.log(n)}
+
+// const addNums = (n1, n2, logAddition) => {
+//     let added = n1 + n2;
+//     logAddition(added);
+// }
+
+// addNums(5, 10, logNumber);
+
+
 
 // Why do we use promises?
 
-// 9.
+// 9. We use promises when we want to perform a task that can take an unknown amount of time to process. When using promises to perform these asynchronous actions, we can temporarily skip passed the promise and perform the next code in the file while we wait for the promise to get data or do something. By using promises we can decrease the amount of time used to execute the code, maximising efficiency
+
 
 // What are the two methods at our disposal if we have hold of a promise? What are the names of the functions that are related to these functions (that are called when the promise is fulfilled or fails).
 
-// 10.
+// 10. If we have hold of a promise we can use .then to get the response object, or .catch to handle any erros that arise
```

Comment:
and inside of the promise (if we're defining our own promise) we have `resolve` and `reject`